postgis/flatgeobuf.c

```diff
@@ -0,0 +1,1039 @@
+
+#include <math.h>
+#include "flatgeobuf.h"
+#include "pgsql_compat.h"
+#include "funcapi.h"
+#include "parser/parse_type.h"
+#include "utils/jsonb.h"
+
+uint8_t magicbytes[] = { 0x66, 0x67, 0x62, 0x03, 0x66, 0x67, 0x62, 0x00 };
+uint8_t MAGICBYTES_LEN = (sizeof(magicbytes) / sizeof((magicbytes)[0]));
+
+static GeometryType_enum_t get_geometrytype(LWGEOM *lwgeom) {
+	int type = lwgeom->type;
+	switch (type)
+	{
+	case POINTTYPE:
+		return GeometryType_Point;
+	case LINETYPE:
+		return GeometryType_LineString;
+	case TRIANGLETYPE:
+		return GeometryType_Triangle;
+	case POLYGONTYPE:
+		return GeometryType_Polygon;
+	case MULTIPOINTTYPE:
+		return GeometryType_MultiPoint;
+	case MULTILINETYPE:
+		return GeometryType_MultiLineString;
+	case MULTIPOLYGONTYPE:
+		return GeometryType_MultiPolygon;
+	case TINTYPE:
+	case COLLECTIONTYPE:
+		return GeometryType_GeometryCollection;
+	default:
+		elog(ERROR, "get_geometrytype: '%s' geometry type not supported",
+				lwtype_name(type));
+	}
+	return GeometryType_Unknown;
+}
+
+static ColumnType_enum_t get_column_type(Oid typoid) {
+	switch (typoid)
+	{
+	case BOOLOID:
+		return ColumnType_Bool;
+	case INT2OID:
+		return ColumnType_Short;
+	case INT4OID:
+		return ColumnType_Int;
+	case INT8OID:
+		return ColumnType_Long;
+	case FLOAT4OID:
+		return ColumnType_Float;
+	case FLOAT8OID:
+		return ColumnType_Double;
+	case TEXTOID:
+	case VARCHAROID:
+		return ColumnType_String;
+	/*case JSONBOID:
+		return ColumnType_Json;*/
+	case BYTEAOID:
+		return ColumnType_Binary;
+	case DATEOID:
+	case TIMEOID:
+	case TIMESTAMPOID:
+	case TIMESTAMPTZOID:
+		return ColumnType_DateTime;
+	}
+	elog(ERROR, "get_column_type: '%d' column type not supported",
+		typoid);
+}
+
+static void encode_header(struct flatgeobuf_encode_ctx *ctx)
+{
+	size_t size;
+	uint8_t *header;
+	flatcc_builder_t builder, *B;
+	uint32_t columns_len = 0;
+	Oid tupType = HeapTupleHeaderGetTypeId(ctx->row);
+	int32 tupTypmod = HeapTupleHeaderGetTypMod(ctx->row);
+	TupleDesc tupdesc = lookup_rowtype_tupdesc(tupType, tupTypmod);
+	int natts = tupdesc->natts;
+	bool geom_found = false;
+	Column_ref_t *columns = lwalloc(sizeof(Column_ref_t) * natts);
+
+	ctx->tupdesc = tupdesc;
+
+	// inspect first geometry
+	if (ctx->lwgeom != NULL) {
+		ctx->hasZ = lwgeom_has_z(ctx->lwgeom);
+		ctx->hasM = lwgeom_has_m(ctx->lwgeom);
+		ctx->lwgeom_type = ctx->lwgeom->type;
+		ctx->geometry_type = get_geometrytype(ctx->lwgeom);
+	} else {
+		ctx->geometry_type = GeometryType_Unknown;
+	}
+
+	B = &builder;
+	flatcc_builder_init(B);
+
+	// inspect columns
+	// NOTE: last element will be unused if geom attr is found
+	for (int i = 0; i < natts; i++) {
+		Oid typoid = getBaseType(TupleDescAttr(tupdesc, i)->atttypid);
+		const char *key = TupleDescAttr(tupdesc, i)->attname.data;
+		POSTGIS_DEBUGF(3, "flatgeobuf: inspecting column definition for %s with oid %d", key, typoid);
+		if (ctx->geom_name == NULL) {
+			if (!geom_found && typoid == postgis_oid(GEOMETRYOID)) {
```

Comment:
If you use postgis_oid, you need to ensure the caller with access to fcinfo calls `postgis_initialize_cache(fcinfo);`. In this case I believe (and I'm using Github as GUI so I'm likely to be wrong) that would be `pgis_asflatgeobuf_finalfn`. See a recent commit in https://github.com/postgis/postgis/commit/678f526670b6b1805495431e62f12072bb9485cf for examples that require this same fix.

I'll have a look at the whole PR once I can fetch the branch locally and see it in my local environment.